src/v/storage/tests/storage_e2e_test.cc

```diff
@@ -2892,6 +2893,166 @@ FIXTURE_TEST(write_truncate_compact, storage_test_fixture) {
+    overrides.cleanup_policy_bitflags
+      = model::cleanup_policy_bitflags::compaction;
+
+    auto ntp = model::ntp("default", "test", 0);
+    auto log = mgr
+                 .manage(
+                   storage::ntp_config(
+                     ntp,
+                     mgr.config().base_dir,
+                     std::make_unique<storage::ntp_config::default_overrides>(
+                       overrides)),
+                   raft::group_id{0},
+                   model::offset_translator_batch_types())
+                 .get();
+    log->start(std::nullopt).get();
+
+    storage::log_append_config appender_cfg{
+      .should_fsync = storage::log_append_config::fsync::no,
+      .io_priority = ss::default_priority_class(),
+      .timeout = model::no_timeout,
+    };
+
+    auto append_random_batch = [&](model::record_batch_type bt) {
+        storage::record_batch_builder builder(bt, model::offset(0));
+        auto num_records = random_generators::get_int(1, 5);
+        for (int i = 0; i < num_records; i++) {
+            builder.add_raw_kv(
+              reflection::to_iobuf(ssx::sformat("key-{}", i)),
+              bytes_to_iobuf(random_generators::get_bytes(5)));
+        }
+        auto reader = model::make_memory_record_batch_reader(
+          std::move(builder).build());
+        std::move(reader)
+          .for_each_ref(log->make_appender(appender_cfg), model::no_timeout)
+          .discard_result()
+          .then([&log] { return log->flush(); })
+          .get();
+    };
+
+    // populate some data, load duplicates of each batch type and
+    // sprinkle some raft batches around.
+    using enum_int_t = std::underlying_type_t<model::record_batch_type>;
+    auto max = static_cast<enum_int_t>(model::record_batch_type::MAX);
+    for (enum_int_t type = 2; type <= max; type++) {
+        auto bt = static_cast<model::record_batch_type>(type);
+        append_random_batch(bt);
+        if (tests::random_bool()) {
+            // some interleaved data batches
+            append_random_batch(model::record_batch_type::raft_data);
+        }
+        append_random_batch(bt);
+    }
+
+    print_batch_info(log);
+
+    // Save the translation map for comparison after compaction.
+    chunked_hash_map<model::offset, model::offset> log_to_kafka;
+    chunked_hash_map<model::offset, model::offset> kafka_to_log;
+    model::offset begin{0};
+    for (model::offset o = begin; o < log->offsets().dirty_offset;
+         o = model::next_offset(o)) {
+        log_to_kafka[o] = log->from_log_offset(o);
+        kafka_to_log[o] = log->to_log_offset(o);
+    }
+
+    // compact the log
+    log->force_roll(ss::default_priority_class()).get();
+    storage::housekeeping_config compaction_cfg(
+      model::timestamp::min(),
+      std::nullopt,
+      model::offset::max(),
+      std::nullopt,
+      ss::default_priority_class(),
+      as);
+    log->housekeeping(compaction_cfg).get();
```

Comment:
does part of the test ensure that a round of compaction completed and some of the data was actually compacted?

Reply:
I verified it manually but I can throw a check in here to ensure that.

Reply:
no big deal i guess i'm just concerned that its non-deterministic